Sources/AsyncKit/ConnectionPool.swift

```diff
@@ -216,23 +249,41 @@ public final class ConnectionPool<Source> where Source: ConnectionPoolSource  {
     /// Connection pools must be closed before they deinitialize.
     ///
     /// - returns: A future indicating close completion.
-    public func close() -> EventLoopFuture<Void> {
-        self.isClosed = true
-        return self.available.map { $0.close() }.flatten(on: self.eventLoop).map {
-            // inform any waiters that they will never be receiving a connection
-            while let waiter = self.waiters.popFirst() {
-                waiter.fail(ConnectionPoolError.closed)
+    public func shutdown() {
+        // synchronize access to closing
+        self.lock.lock()
+
+        // check to make sure we aren't double closing
+        guard !self.didShutdown else {
+            return
+        }
+        self.didShutdown = true
+
+        // is closed variable mutated
+        self.lock.unlock()
+
+        // no locks needed as this can only happen once
+        for available in self.available {
```

Comment:
I think this could still race with makeConnection and/or requestConnection, why not just keep the log around for longer to be safe? (Also see below)